src/forest.cpp

```diff
@@ -341,5 +342,533 @@ Set the parent and edge label for a node. This function sets the parent of
+Range object for iterating through paths in a :any:`Forest`.
+
+This class represents a range object that facilitates
+iterating through the paths from every node to the root of its subtree in a
+:any:`Forest` object. These nodes are taken in numerical order, so the first value
+returned by :any:`get` is the word to a root from node ``0``, after :any:`next`
+is called, :any:`get` returns the word to a root from node ``1``, and so on.
+The point of this class is to permit more efficient iteration over many paths
+in a :any:`Forest` object than :any:`path_to_root` (and its
+associated helper functions). :any:`path_to_root` traverses
+the :any:`Forest` from the given node to the root of its tree. If the path from
+nodes ``m`` and ``n`` to the root of their tree share a long common suffix, then
+this suffix will be traversed twice in two calls to :any:`path_to_root`.
+:any:`PathsToRoots` avoids this by finding the least common ancestor of ``m`` and
+``n``, so that the suffix is not re-traversed. This works best when the nodes in
+the :any:`Forest` are specified in a sensible order (such as via a depth-first or
+breadth-first traversal).)pbdoc");
+
+      thing.def("__repr__", [](forest::PathsToRoots const& thing) {
+        return to_human_readable_repr(thing, ".");
+      });
+
+      thing.def("__copy__", [](forest::PathsToRoots const& pfr) {
+        return forest::PathsToRoots(pfr);
+      });
+
+      thing.def(py::init<Forest const&>(),
+                py::arg("f"),
+                R"pbdoc(
+:sig=(self: PathsToRoots, f: Forest) -> None:
+
+Construct from a :any:`Forest`.
+
+This function constructs a new :any:`PathsToRoots` for the :any:`Forest` *f*.
+
+The newly constructed object does not copy *f* and is not valid if *f* is
+destroyed.
+
+:param f: the Forest.
+:type f: Forest
+)pbdoc");
+
+      thing.def(
+          "init",
+          [](forest::PathsToRoots& self,
+             Forest const&         f) -> forest::PathsToRoots& {
+            return static_cast<forest::PathsToRoots&>(self.init(f));
+          },
+          py::arg("f"),
+          R"pbdoc(
+:sig=(self: PathsToRoots, f: Forest) -> PathsToRoots:
+
+Re-initialize from a :any:`Forest`.
+
+This function re-initializes *self* so that its underlying :any:`Forest`
+object is *f*. This puts *self* back into the same state it
+would have been in if it had been newly constructed from *f*.
+
+:param f: the Forest.
+:type f: Forest
+
+:returns: ``self``
+:rtype: PathsToRoots
+)pbdoc");
+
+      thing.def(
+          "copy",
+          [](forest::PathsToRoots const& pfr) {
+            return forest::PathsToRoots(pfr);
+          },
+          R"pbdoc(
+:sig=(self: PathsToRoots) -> PathsToRoots:
+
+Copy a :any:`PathsToRoots` object.
+
+:returns: A copy.
+:rtype: PathsToRoots
+)pbdoc");
+
+      thing.def(
+          "__iter__",
+          [](forest::PathsToRoots& self) {
+            return py::make_iterator(self.begin(), self.end());
+          },
+          py::keep_alive<0, 1>(),
```

Comment:
I would benefit from a comment that explains what this slightly cryptic thing is doing, but maybe I'm just being a noob. My understanding is that this means "Keep `this` alive for at least as long as the returned iterator is alive"?

Feel free to disregard if you think this is clear as it is.